### PR TITLE
Add relay encoder

### PR DESCRIPTION
**Pull request by DZahar0v**

### CODE REVIEW

File: contracts/RelayEncoder.sol [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            value = value >> 8;
+        }
+        return out;
+    }
+
+    /**
+    * @notice Converting uint256 value to bytes
+    * @param value - uint256 value
+    */
+    function scaleCompactUint(uint256 value) internal pure returns(bytes memory) {
```
Comment (algys) [inline contracts/RelayEncoder.sol:100]: I think we can move that piece of code to lib

File: contracts/RelayEncoder.sol [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+    // dev Encode 'withdraw_unbonded' relay call
+    // @param slashes: Weight hint, number of slashing spans
+    // @returns The bytes associated with the encoded call
+    function encode_withdraw_unbonded(uint32 slashes) external pure returns (bytes memory) {
+        if (slashes < 1<<8) {
+            return bytes.concat(hex"0603", bytes1(uint8(slashes)), bytes3(0));
+        }
+        if(slashes < 1 << 16) {
+            uint32 bt2 = slashes / 256;
```
Comment (algys) [inline contracts/RelayEncoder.sol:44]: bitwise shift should be lighter ?  slashes << 8

File: contracts/RelayEncoder.sol [hunk 3]
```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+contract RelayEncoder { // NOTE: this is logic for proxy contract
```
Comment (algys) [inline contracts/RelayEncoder.sol:4]: proxy ? why ?

**Review by algys — APPROVED** [review]

File: contracts/RelayEncoder.sol (hunk 3 above)
Comment (algys) [reply contracts/RelayEncoder.sol:4]: also you can derive from interface

File: contracts/RelayEncoder.sol [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    function encode_chill() external pure returns (bytes memory) {
+        return hex"0606";
+    }
+
+    /**
+    * @notice Converting uint256 value to le bytes
+    * @param value - uint256 value
+    * @param len - length of output bytes array
+    */
+    function toLeBytes(uint256 value, uint256 len) internal pure returns(bytes memory) {
```
Comment (algys) [inline contracts/RelayEncoder.sol:87]: this one might be moved to lib as well

**Review by algys — CHANGES_REQUESTED** [review]

File: contracts/RelayEncoder.sol [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+pragma solidity ^0.8.0;
+
+contract RelayEncoder { // NOTE: this is logic for proxy contract
+
+    // dev Encode 'bond' relay call
+    // @param controller_address: Address of the controller
+    // @param amount: The amount to bond
+    // @param reward_destination: the account that should receive the reward
+    // @returns The bytes associated with the encoded call
+    function encode_bond(uint256 controller_address, uint256 amount, bytes memory reward_destination) external pure returns (bytes memory result) {
```
[the remark this answers is not in the document]
Comment (DZahar0v) [reply contracts/RelayEncoder.sol:11]: Fixed

File: contracts/RelayEncoder.sol [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        return bytes.concat(hex"0600", hex"00", bytes32(controller_address), scaleCompactUint(amount), reward_destination);
+    }
+
+    // dev Encode 'bond_extra' relay call
+    // @param amount: The extra amount to bond
+    // @returns The bytes associated with the encoded call
+    function encode_bond_extra(uint256 amount) external pure returns (bytes memory) {
+        return bytes.concat(hex"0601", scaleCompactUint(amount));
```
[the remark this answers is not in the document]
Comment (DZahar0v) [reply contracts/RelayEncoder.sol:19]: Done

File: contracts/RelayEncoder.sol (hunk 1 above)
Comment (DZahar0v) [reply contracts/RelayEncoder.sol:100]: Done

File: contracts/RelayEncoder.sol (hunk 3 above)
Comment (DZahar0v) [reply contracts/RelayEncoder.sol:4]: Done

File: contracts/RelayEncoder.sol (hunk 4 above)
Comment (DZahar0v) [reply contracts/RelayEncoder.sol:87]: Done

File: contracts/RelayEncoder.sol (hunk 2 above)
Comment (DZahar0v) [reply contracts/RelayEncoder.sol:44]: Not Fixed

**Review by algys — APPROVED** [review]

**Pull request merged**